File: Controllers/AccountController.vb
Imports System.Web.Mvc
Imports System.Web.Security
Imports My.Models
Imports My.Helpers

Namespace My.Controllers
    Public Class AccountController
        Inherits Controller

        ' GET: /Account/Login
        Function Login(returnUrl As String) As ActionResult
            ViewBag.ReturnUrl = returnUrl
            Return View()
        End Function

        ' POST: /Account/Login
        <HttpPost>
        <ValidateAntiForgeryToken>
        Function Login(model As LoginViewModel, returnUrl As String) As ActionResult
            If Not ModelState.IsValid Then Return View(model)

            ' TODO: 查詢資料庫驗證帳號密碼
            ' Dim encPass = CryptoHelper.Encrypt(model.Password)
            ' Dim member = db.Members.FirstOrDefault(Function(m) m.Account = model.Account AndAlso m.Password = encPass)
            Dim member As Member = Nothing ' 暫定，待接資料庫

            If member Is Nothing Then
                ModelState.AddModelError("", "帳號或密碼錯誤")
                Return View(model)
            End If

            FormsAuthentication.SetAuthCookie(member.Account, model.RememberMe)
            Session("MemberId") = member.Id
            Session("MemberName") = member.Name

            If Not String.IsNullOrEmpty(returnUrl) AndAlso Url.IsLocalUrl(returnUrl) Then
                Return Redirect(returnUrl)
            End If
            Return RedirectToAction("Index", "Home")
        End Function

        ' GET: /Account/Register
        Function Register() As ActionResult
            Return View()
        End Function

        ' POST: /Account/Register
        <HttpPost>
        <ValidateAntiForgeryToken>
        Function Register(model As RegisterViewModel) As ActionResult
            If Not ModelState.IsValid Then Return View(model)

            ' TODO: 檢查帳號是否重複，寫入資料庫
            ' Dim exists = db.Members.Any(Function(m) m.Account = model.Account)
            ' If exists Then ...

            ViewBag.Message = "註冊成功，請登入！"
            Return RedirectToAction("Login")
        End Function

        ' GET: /Account/Logout
        Function Logout() As ActionResult
            FormsAuthentication.SignOut()
            Session.Clear()
            Return RedirectToAction("Index", "Home")
        End Function

        ' GET: /Account/Profile
        <Authorize>
        Function Profile() As ActionResult
            ' TODO: 讀取會員資料
            Return View()
        End Function

        ' GET: /Account/ChangePassword
        <Authorize>
        Function ChangePassword() As ActionResult
            Return View()
        End Function

        ' POST: /Account/ChangePassword
        <HttpPost>
        <Authorize>
        <ValidateAntiForgeryToken>
        Function ChangePassword(model As ChangePasswordViewModel) As ActionResult
            If Not ModelState.IsValid Then Return View(model)
            ' TODO: 驗證舊密碼並更新新密碼
            ViewBag.Message = "密碼已成功更新！"
            Return View()
        End Function

        ' GET: /Account/ForgetPassword
        Function ForgetPassword() As ActionResult
            Return View()
        End Function

        ' POST: /Account/ForgetPassword
        <HttpPost>
        <ValidateAntiForgeryToken>
        Function ForgetPassword(email As String) As ActionResult
            ' TODO: 寄送重設密碼 Email
            ViewBag.Message = "重設密碼連結已寄至您的信箱"
            Return View()
        End Function

    End Class
End Namespace

File: Controllers/AccountController.cs
using System.Web.Mvc;
using System.Web.Security;
using MyWeb.Models;
using MyWeb.Helpers;

namespace MyWeb.Controllers
{
    public class AccountController : Controller
    {
        // GET: /Account/Login
        public ActionResult Login(string returnUrl)
        {
            ViewBag.ReturnUrl = returnUrl;
            return View();
        }

        // POST: /Account/Login
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Login(LoginViewModel model, string returnUrl)
        {
            if (!ModelState.IsValid) return View(model);

            // TODO: 查詢資料庫驗證帳號密碼
            // var member = db.Members.FirstOrDefault(m => m.Account == model.Account);
            // if (member == null || !CryptoHelper.Verify(model.Password, member.Password)) ...
            Member member = null; // 暫定，待接資料庫

            if (member == null)
            {
                ModelState.AddModelError("", "帳號或密碼錯誤");
                return View(model);
            }

            FormsAuthentication.SetAuthCookie(member.Account, model.RememberMe);
            Session["MemberId"] = member.Id;
            Session["MemberName"] = member.Name;

            if (!string.IsNullOrEmpty(returnUrl) && Url.IsLocalUrl(returnUrl))
            {
                return Redirect(returnUrl);
            }
            return RedirectToAction("Index", "Home");
        }

        // GET: /Account/Register
        public ActionResult Register()
        {
            return View();
        }

        // POST: /Account/Register
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Register(RegisterViewModel model)
        {
            if (!ModelState.IsValid) return View(model);

            // TODO: 檢查帳號是否重複，寫入資料庫
            // var exists = db.Members.Any(m => m.Account == model.Account);
            // if (exists) ...

            ViewBag.Message = "註冊成功，請登入！";
            return RedirectToAction("Login");
        }

        // GET: /Account/Logout
        public ActionResult Logout()
        {
            FormsAuthentication.SignOut();
            Session.Clear();
            return RedirectToAction("Index", "Home");
        }

        // GET: /Account/Profile
        [Authorize]
        public ActionResult Profile()
        {
            // TODO: 讀取會員資料
            return View();
        }

        // GET: /Account/ChangePassword
        [Authorize]
        public ActionResult ChangePassword()
        {
            return View();
        }

        // POST: /Account/ChangePassword
        [HttpPost]
        [Authorize]
        [ValidateAntiForgeryToken]
        public ActionResult ChangePassword(ChangePasswordViewModel model)
        {
            if (!ModelState.IsValid) return View(model);
            // TODO: 驗證舊密碼並更新新密碼
            ViewBag.Message = "密碼已成功更新！";
            return View();
        }

        // GET: /Account/ForgetPassword
        public ActionResult ForgetPassword()
        {
            return View();
        }

        // POST: /Account/ForgetPassword
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult ForgetPassword(string email)
        {
            // TODO: 寄送重設密碼 Email
            ViewBag.Message = "重設密碼連結已寄至您的信箱";
            return View();
        }
    }
}
